### PR TITLE
clusters general validation was covered with tests

### PR DESCRIPTION
**Pull request by tengu-alt**

### CODE REVIEW

File: apis/clusters/v1beta1/validation_test.go [hunk 1]
```diff
@@ -0,0 +1,1063 @@
...
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if err := validateTagsUpdate(tt.args.new, tt.args.old); (err != nil) != tt.wantErr {
+				t.Errorf("validateTagsUpdate() error = %v, wantErr %v", err, tt.wantErr)
+			}
+		})
+	}
+}
```
Comment (DoodgeMatvey) [inline apis/clusters/v1beta1/validation_test.go:519]: Why do you have only error cases?
Could you please also add successful cases where there are none?

**Review by DoodgeMatvey — CHANGES_REQUESTED** [review]

File: apis/clusters/v1beta1/validation_test.go (hunk 1 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/validation_test.go:519]: it has a lot of cases when we use valid data so there are successful cases also.Als

Comment (DoodgeMatvey) [reply apis/clusters/v1beta1/validation_test.go:519]: What do you mean? You don't have any `wantErr: false` for `Test_validateTagsUpdate` or `Test_validateClusterManagedNodes` and there are other functions

Comment (tengu-alt) [reply apis/clusters/v1beta1/validation_test.go:519]: I have `wantErr: false` in cases that validate creation on updating of resource such as `Test_validateTwoFactorDelete` or `Test_validateAppVersion`. But do we need a success cases when data is immutable?(`old` and  `new` structs are identical)

Comment (DoodgeMatvey) [reply apis/clusters/v1beta1/validation_test.go:519]: Yea, in tests you need to handle error and successful cases by best practices, so I think it makes sense.

File: apis/clusters/v1beta1/validation.go [hunk 2]
```diff
@@ -172,7 +168,7 @@ func (cps *CloudProviderSettings) ValidateCreation() error {
 
 func validateReplicationFactor(availableReplicationFactors []int, rf int) error {
 	if rf <= 0 || rf > 300 {
-		return fmt.Errorf("replication factor must be a multiple of %v, up to a maximum value of 300",
+		return fmt.Errorf("replication factor must be a equal of %v, up to a maximum value of 300",
```
Comment (testisnullus) [inline apis/clusters/v1beta1/validation.go:171]: a -> an

File: apis/clusters/v1beta1/validation.go [hunk 3]
```diff
@@ -182,7 +178,7 @@ func validateReplicationFactor(availableReplicationFactors []int, rf int) error
 		}
 	}
 
-	return fmt.Errorf("replication factor must be a multiple of %v, up to a maximum value of 300",
+	return fmt.Errorf("replication factor must be a equal of %v, up to a maximum value of 300",
```
Comment (testisnullus) [inline apis/clusters/v1beta1/validation.go:181]: same is here

File: apis/clusters/v1beta1/validation_test.go [hunk 4]
```diff
@@ -0,0 +1,1072 @@
...
+
+	"github.com/instaclustr/operator/pkg/models"
+)
+
+func TestCluster_ValidateCreation(t *testing.T) {
+	type fields struct {
+		Name                  string
+		Version               string
+		PCICompliance         bool
+		PrivateNetworkCluster bool
```
Comment (testisnullus) [inline apis/clusters/v1beta1/validation_test.go:14]: PrivateNetworkCluster is named like PrivateNetwork inside the `GenericClusterSpec` for now

File: apis/clusters/v1beta1/validation_test.go [hunk 5]
```diff
@@ -0,0 +1,1072 @@
...
+	}
+}
+
+func TestDataCentre_ValidateCreation(t *testing.T) {
+	type fields struct {
+		Name                  string
+		Region                string
+		CloudProvider         string
+		ProviderAccountName   string
+		CloudProviderSettings []*CloudProviderSettings
```
Comment (testisnullus) [inline apis/clusters/v1beta1/validation_test.go:88]: There are no `CloudProviderSettings`, `NodeSize`, and `NodesNumber` fields inside the GenericDataCentreSpec. Could you clarify which specific aspects of the cluster specification validation are being tested in this pull request?

**Review by testisnullus — CHANGES_REQUESTED** [review]

File: apis/clusters/v1beta1/validation.go (hunk 2 above)
Comment (worryg0d) [reply apis/clusters/v1beta1/validation.go:171]: may be `replication factor must be one of %v....`, wdyt?

File: apis/clusters/v1beta1/validation_test.go (hunk 5 above)
Comment (worryg0d) [reply apis/clusters/v1beta1/validation_test.go:88]: The same case as we have above with `Cluster`

File: apis/clusters/v1beta1/validation_test.go (hunk 1 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/validation_test.go]: agree, refactored

File: apis/clusters/v1beta1/validation.go (hunk 2 above)
Comment (tengu-alt) [reply apis/clusters/v1beta1/validation.go]: agree changed to ` must be one of %v`

**Review by DoodgeMatvey — APPROVED** [review]

File: apis/clusters/v1beta1/validation.go [hunk 6]
```diff
@@ -172,7 +168,7 @@ func (cps *CloudProviderSettings) ValidateCreation() error {
 
 func validateReplicationFactor(availableReplicationFactors []int, rf int) error {
 	if rf <= 0 || rf > 300 {
-		return fmt.Errorf("replication factor must be a multiple of %v, up to a maximum value of 300",
+		return fmt.Errorf("replication factor must be one of %v, up to a maximum value of 300",
 			availableReplicationFactors)
 	}
```
Comment (ribaraka) [inline apis/clusters/v1beta1/validation.go:173]: Is 300 the maximum permissible value for the replica factor?

**Review by ribaraka — APPROVED** [review]

**Review by testisnullus — APPROVED** [review]

**Pull request merged**